### DISCUSSION TITLE
Azure DevOps - how can we get the integer build definition ID?

### DISCUSSION THREAD

**Discussion by JohnLudlow** [pulumi/pulumi#12647] [Pulumi Providers]
Hi,

I have a number of Azure DevOps git repos with corresponding pipelines (build definitions). I'd like to apply a policy that validates the PR build passes before allowing a PR to complete.

It seems as if `BranchPolicyBuildValidation.Settings.BuildDefintionId` expects an int (which makes sense because there is a numerical definition Id) but every way I've found so far of getting a build definition (e.g `GetBuildDefinition`) has a string property `Id` which is the name of the definition (e.g "MySpecialBuildDefinition") not the int id (e.g "1234").

```csharp
var azureBuildDefinition = GetBuildDefinition.Invoke(new GetBuildDefinitionInvokeArgs() { 
    ProjectId = args.Project.Apply(p => p.Id),
                
    Name = args.Repository.Name,
});

var azureRepository = GetRepositories.Invoke(new GetRepositoriesInvokeArgs()
{
    ProjectId = args.Project.Apply(p => p.Id),

    Name = args.Repository.Name,
});

_ = new BranchPolicyBuildValidation($"{args.Repository.Name}-branchPolicyBuildValidation", new()
{
    ProjectId = args.Project.Apply(p => p.Id),
                
    Enabled = true,
    Blocking = true,
    Settings = new Pulumi.AzureDevOps.Inputs.BranchPolicyBuildValidationSettingsArgs() {

        ///////
        // This is the problem line
        ///////
        BuildDefinitionId = azureBuildDefinition.Apply(b => b.Id),

        DisplayName = $"{args.Repository.Name} / {azureBuildDefinition.Apply(b => b.Name)} Build Validation policy",
        ValidDuration = 1440,

        Scopes = new[]
        {
            new Pulumi.AzureDevOps.Inputs.BranchPolicyBuildValidationSettingsScopeArgs
            {
                RepositoryId = azureRepository.Apply(r => r.Id),
                MatchType = "DefaultBranch"
            }
        }
    }
});
```

The code above fails to compile because the types don't match when setting `BuildDefinitionId = azureBuildDefinition.Apply(b => b.Id)`, and it's not just a string that contains a number either (so `int.Parse` doesn't work here).

The [documentation](https://www.pulumi.com/registry/packages/azuredevops/api-docs/branchpolicybuildvalidation/) shows setting up a new repo and using its output:

```csharp
    var exampleBuildDefinition = new AzureDevOps.BuildDefinition("exampleBuildDefinition", new()
    {
        ProjectId = exampleProject.Id,
        Repository = new AzureDevOps.Inputs.BuildDefinitionRepositoryArgs
        {
            RepoType = "TfsGit",
            RepoId = exampleGit.Id,
            YmlPath = "azure-pipelines.yml",
        },
    });

    var exampleBranchPolicyBuildValidation = new AzureDevOps.BranchPolicyBuildValidation("exampleBranchPolicyBuildValidation", new()
    {
        ProjectId = exampleProject.Id,
        Enabled = true,
        Blocking = true,
        Settings = new AzureDevOps.Inputs.BranchPolicyBuildValidationSettingsArgs
        {
            DisplayName = "Example build validation policy",
            BuildDefinitionId = exampleBuildDefinition.Id,
            ValidDuration = 720
        }   
    }
```

This seems to fail for the same reason - `exampleBuildDefinition.Id` is a string, and its content is not numeric.

Am I missing an obvious way to get the id of the definition?

**Comment by JohnLudlow** [comment]
Oh, I have Pulumi `v3.61.1`

**Comment by JohnLudlow** [comment]
Ok never mind - I was reading the error wrong and it was actually complaining about something else which I have now resolved

**Comment by dirien** [comment]
Hi @JohnLudlow,

good that you found the issue! Any further questions, feel free to open a new thread or join the community slack. https://slack.pulumi.com

**Comment by JohnLudlow** [comment]
Will do, thanks!
